R/csv-to-duckdb.R

```diff
@@ -0,0 +1,303 @@
+      ")"
+    )
+  }
+
+  paste0(
+    sprintf("%s AS %s", old_column_names, new_column_names),
+    collapse = ",\n  "
+  )
+}
+
+make_sql_schema_string <- function(con, schema, dt_override = NULL) {
+  stopifnot(is.character(schema))
+  stopifnot(!is.null(names(schema)))
+  check_duckdb_types(con, schema)
+
+  if (!is.null(dt_override)) {
+    # import special date/datetime columns as VARCHAR because they will be
+    # converted in SELECT
+    schema[names(dt_override)] <- "VARCHAR"
+  }
+
+  paste0(
+    "{\n",
+    paste(
+      sprintf("    '%s': '%s'", names(schema), schema),
+      collapse = ",\n"
+    ),
+    "\n  }"
+  )
+}
+
+check_duckdb_types <- function(con, schema) {
+  valid_types <- DBI::dbGetQuery(
+    con,
+    "SELECT type_name FROM duckdb_types()"
+  )[["type_name"]]
+
+  not_in <- setdiff(tolower(schema), valid_types)
+
+  if (length(not_in) > 0) {
+    stop(
+      "The following are not valid duckdb data types: '",
+      paste0(not_in, collapse = "', '"),
+      "'"
+    )
+  }
+}
+
+#' Sniff a csv to get duckdb's auto-detected spec for the file
+#'
+#' This is a good function to run before [csv_to_duckdb_query()] to get
+#' a starting point, especially for the schema.
+#'
+#' @inheritParams csv_to_duckdb_query
+#' @param sample_size How many rows to sample to sniff formats from the data
+#'
+#' @return a list, with auto-detected `date_format`, `timestamp_format`, and
+#'    `schema`
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' spec <- sniff_csv_duckdb("path_to_csv.csv")
+#'
+#' # Use dput to get a schema you can edit
+#' dput(spec$schema)
+#' }
+sniff_csv_duckdb <- function(path, sample_size = NULL) {
+  con <- DBI::dbConnect(duckdb::duckdb())
+  on.exit(DBI::dbDisconnect(con))
+
+  ss <- ""
+  if (!is.null(sample_size)) {
+    ss <- glue::glue(", sample_size = {sample_size}")
+  }
+
+  res <- DBI::dbGetQuery(
+    con,
+    glue::glue(
+      "SELECT DateFormat, TimestampFormat, Columns FROM sniff_csv('{path}' {ss})"
+    )
+  )
+
+  structure(
+    list(
+      date_format = res$DateFormat,
+      timestamp_format = res$TimestampFormat,
+      schema = stats::setNames(res$Columns[[1]]$type, res$Columns[[1]]$name)
+    ),
+    class = "sniffed_csv"
+  )
+}
+
+#' @export
+print.sniffed_csv <- function(x, ...) {
+  cat("Auto-detected date_format: ", x$date_format)
+  cat("\n")
+  cat("Auto-detected timstamp_format: ", x$timestamp_format)
+  cat("\n")
+  cat("\n")
+
+  cat("Auto-detected schema:\n")
+  cat("  c(\n")
+  cat(paste0("    ", names(x$schema), " = \"", x$schema, collapse = "\",\n"))
+  cat("\"\n  )")
+}
+
+parse_opts <- function(con, nullstr = NULL, opts = list()) {
+  stopifnot(is.list(opts))
+  stopifnot(is.null(nullstr) || is.character(nullstr))
+
+  opts <- remove_nulls(c(list(nullstr = nullstr), opts))
+
+  if (is.null(opts) || length(opts) == 0) {
+    return("")
+  }
+
+  check_params(con, opts)
+
+  varchar_scalar_opts <- Filter(\(x) is.character(x) & length(x) == 1, opts)
```

Comment:
I'm tempted to extract the code that constructs the various *_opts_parsed variables into individual functions.  Consider refactoring into smaller, focused functions like:
`format_varchar_scalar_opts <- function(opts) { ... }`
`format_varchar_array_opts <- function(opts) { ... } `  etc.